typing/printtyp.ml

```diff
@@ -874,28 +878,59 @@ end = struct
     if non_gen then new_weak_name ty ()
     else new_name ()
 
-  let name_of_type name_generator t =
+  let name_of_type ~non_gen name_generator t =
     (* We've already been through repr at this stage, so t is our representative
        of the union-find class. *)
     let t = substitute t in
     try List.assq t !names with Not_found ->
       try TransientTypeMap.find t !weak_var_map with Not_found ->
       let name =
         match t.desc with
-          Tvar (Some name) | Tunivar (Some name) ->
-            (* Some part of the type we've already printed has assigned another
-             * unification variable to that name. We want to keep the name, so
-             * try adding a number until we find a name that's not taken. *)
+          Tvar (Some name) | Tunivar (Some name) when not non_gen ->
             let available name =
               List.for_all
                 (fun (_, name') -> name <> name')
                 !names
             in
             if available name then name
             else
-              let suffixed i = name ^ Int.to_string i in
-              let i = Misc.find_first_mono (fun i -> available (suffixed i)) in
-              suffixed i
+              (* Some part of the type we've already printed has assigned
+                 another unification variable to that name. Yet we want to
+                 keep the name we have, even though it's taken. Our approach:
+                 - If the name is a single letter, try the next 4 single
+                   letters. This will get us from 'a to 'e or from 'k to 'o.
+                   Why 4? No good reason; it just seems the right number.
+                 - Otherwise, add an increasing numeric suffix until we find an
+                   available name.
+              *)
+              let with_suffix () =
+                let suffixed i = name ^ Int.to_string i in
+                let i =
+                  Misc.find_first_mono (fun i -> available (suffixed i))
+                in
+                suffixed i
+              in
+              if String.length name = 1 &&
+                 Char.compare name.[0] 'a' >= 0 &&
+                 Char.compare name.[0] 'z' <= 0
+              then
+                let code = Char.code name.[0] in
+                let max_num_letters_to_try = 4 in
+                let num_letters = 26 in
+                let num_letters_left =
+                  num_letters - (code - Char.code 'a') - 1
+                in
+                let actual_num_letters_to_try =
+                  Int.min max_num_letters_to_try num_letters_left
+                in
+                let possible_names =
+                  List.init actual_num_letters_to_try
+                    (fun n -> String.make 1 (Char.chr (code + n + 1)))
```

Comment:
Nitpick: I find it slightly messy that the code mixes higher-level concerns (weak_var_map, available, !names, etc.) with low-level concerns (character/string manipulation). I would prefer if the character/string stuff was moved to an auxiliary function, I think the code would read nicer. I see two reasonable ways to do this:

- The entirety of the `else` branch of `if available name then name else ...` could move to an auxiliary function, for example `pick_good_name available name`. The nice comment you wrote would document what this function does.
- Or maybe you could have a function `suggest_good_names : string -> string list` that only does the non-fallback part (computes the at-most-four suggestions), with the try-or-fallback logic still in the function.